hogql_parser/parser.cpp

```diff
@@ -291,6 +291,447 @@ class HogQLParseTreeConverter : public HogQLParserBaseVisitor {
+      }
+    } else if (increment_expression_ctx) {
+      try {
+        increment = visitAsPyObject(increment_expression_ctx);
+      } catch (...) {
+        Py_DECREF(initializer);
+        Py_DECREF(condition);
+        throw;
+      }
+    } else {
+      increment = Py_None;
+      Py_INCREF(increment);
+    }
+
+    PyObject* body;
+    try {
+      body = visitAsPyObject(ctx->statement());
+    } catch (...) {
+      Py_DECREF(initializer);
+      Py_DECREF(condition);
+      Py_DECREF(increment);
+      throw;
+    }
+
+    PyObject* ret = build_ast_node(
+        "ForStatement", "{s:N,s:N,s:N,s:N}", "initializer", initializer, "condition", condition, "increment", increment,
+        "body", body
+    );
+    if (!ret) {
+      Py_DECREF(initializer);
+      Py_DECREF(condition);
+      Py_DECREF(increment);
+      Py_DECREF(body);
+      throw PyInternalError();
+    }
+    return ret;
+  }
+
+  VISIT(FuncStmt) {
+    string name = visitAsString(ctx->identifier());
+    PyObject* params;
+    auto identifier_list_ctx = ctx->identifierList();
+    if (identifier_list_ctx) {
+      try {
+        params = visitAsPyObject(identifier_list_ctx);
+      } catch (...) {
+        throw;
+      }
+    } else {
+      params = PyList_New(0);
+      if (!params) {
+        throw PyInternalError();
+      }
+    }
+
+    PyObject* body;
+    try {
+      body = visitAsPyObject(ctx->block());
+    } catch (...) {
+      Py_DECREF(params);
+      throw;
+    }
+
+    PyObject* ret = build_ast_node("Function", "{s:s#,s:N,s:N}", "name", name.data(), name.size(), "params", params, "body", body);
+    if (!ret) {
+      Py_DECREF(params);
+      Py_DECREF(body);
+      throw PyInternalError();
+    }
+    return ret;
+  }
+
+  VISIT(KvPairList) {
+    PyObject* ret = PyList_New(0);
+    if (!ret) {
+      throw PyInternalError();
+    }
+    auto kv_pair_ctxs = ctx->kvPair();
+    for (auto kv_pair_ctx : kv_pair_ctxs) {
+      PyObject* kv;
+      try {
+        kv = visitAsPyObject(kv_pair_ctx);
+      } catch (...) {
+        Py_DECREF(ret);
+        throw;
+      }
+      int append_code = PyList_Append(ret, kv);
+      Py_DECREF(kv);
+      if (append_code == -1) {
+        Py_DECREF(ret);
+        throw PyInternalError();
+      }
+    }
+    return ret;
+  }
+
+  VISIT(KvPair) {
+    PyObject* k;
+    PyObject* v;
+    try {
+      k = visitAsPyObject(ctx->expression(0));
+      v = visitAsPyObject(ctx->expression(1));
+    } catch (...) {
+      throw;
+    }
+    PyObject* ret = PyTuple_Pack(2, k, v);
+    if (!ret) {
+      Py_DECREF(k);
+      Py_DECREF(v);
+      throw PyInternalError();
+    }
+    return ret;
+  }
+
+  VISIT(IdentifierList) {
+    vector<string> identifiers;
+    auto identifier_ctxs = ctx->identifier();
+    identifiers.reserve(identifier_ctxs.size());
+    for (auto identifier_ctx : identifier_ctxs) {
+      identifiers.push_back(visitAsString(identifier_ctx));
+    }
+    return X_PyList_FromStrings(identifiers);
```

Comment:
I had to revert this. Not sure why, but it failed with `Parsing failed due to bad type casting` (even after adding the `return` I forgot the first time). The above works, so I'll 🙈

Reply:
Though actually this goes against [convention No. 3](https://github.com/PostHog/posthog/blob/540729c7130ab0d72f41d8243390ba07aaab1161/hogql_parser/CONTRIBUTING.md#L42-L43):
> Stay out of Python land as long as possible. E.g. avoid using `PyObject*`s` for bools or strings.

So in this spirit, this should only return the vector of strings, as is conventional. Then the rule that actually builds an AST node using `IdentifierList` can use `X_PyList_FromStrings()` itself – so that this method doesn't have to be needlessly concerned with Python internals